leaderboard/enriching/enricher.go

```diff
@@ -45,34 +55,91 @@ type enricherImpl struct {
 	config EnrichmentConfig
 	logger *zap.Logger
 	client *http.Client
+	cache  EnricherCache
 }
 
 // NewEnricher returns a new Enricher implementation.
-func NewEnricher(config EnrichmentConfig, logger *zap.Logger) Enricher {
+func NewEnricher(
+	config EnrichmentConfig,
+	logger *zap.Logger,
+	cache EnricherCache,
+) Enricher {
 	return &enricherImpl{
 		config: config,
-		logger: logger,
+		logger: logger.With(zap.String("source", "enricher")),
 		client: &http.Client{
 			Timeout: config.WebhookTimeout,
 		},
+		cache: cache,
 	}
 }
 
-// Enrich enriches the members list with some metadata. By default, it will call the Cloud Save service.
-// If there's a webhook configured for the tenantID, it will call it instead.
-func (e *enricherImpl) Enrich(ctx context.Context, tenantID, leaderboardID string, members []*model.Member) ([]*model.Member, error) {
+// Enrich enriches the members list with some metadata.
+// By default, it will call the Cloud Save service, unless it's disabled for the tenantID or if there's a webhook for the tenantID.
+// If there's a webhook configured for the tenantID, it will be called instead.
+func (e *enricherImpl) Enrich(
+	ctx context.Context,
+	tenantID,
+	leaderboardID string,
+	members []*model.Member,
+) ([]*model.Member, error) {
 	if len(members) == 0 {
 		return members, nil
 	}
 
-	tenantUrl, exists := e.config.WebhookUrls[tenantID]
-	if !exists {
+	cached, hit, err := e.cache.Get(ctx, tenantID, leaderboardID, members)
+	if err != nil {
+		e.logger.Error("could not get cached enrichment data", zap.Error(err))
+	}
+
+	if hit {
+		e.logger.Debug("returning cached enrich data")
+		for _, m := range members {
+			if metadata, exists := cached[m.PublicID]; exists {
+				m.Metadata = metadata
+			}
+		}
+
+		return members, nil
+	}
```

Comment:
This piece of code can be wrapped in a decorator called `CacheableEnricher` that implements `Enricher` and receives a implementation of `Enricher`.

First it runs this piece of code and if there is no hit it calls the provided `Enricher` implementation, i.e., the actual implementation, caching the results after.

If you need an example of a decorator, it is something similar to what Pablo has done in Push Notifications Center.

Reply:
I've refactored it. PTAL?